local-cli/server/util/attachHMRServer.js

```diff
@@ -16,11 +16,254 @@ const url = require('url');
+  let clients = [];
+  let clientsPerBundleEntry = {};
+
+  let platformCache = {
+    android: {},
+    ios: {},
+  };
+
+  function addClient (client) {
+    client.ws.on('error', e => {
+      console.error('[Hot Module Replacement] Unexpected error', e);
+      disconnect(client);
+    });
+
+    client.ws.on('close', () => disconnect(client));
+    clients.push(client);
+
+    clientsPerBundleEntry[client.bundleEntry] = [...(
+      clientsPerBundleEntry[client.bundleEntry] ? clientsPerBundleEntry[client.bundleEntry] : []
+    ), client];
+
+    // Only register callback if it's the first client
+    if (clients.length === 1) {
+      packagerServer.setHMRFileChangeListener(onHMRChange);
+    }
+
+  }
+
+  function disconnect (disconnectedClient) {
+    clients = clients.filter(client => client !== disconnectedClient);
+    clientsPerBundleEntry[disconnectedClient.bundleEntry] = clientsPerBundleEntry[disconnectedClient.bundleEntry].filter(client => client !== disconnectedClient);
+    //Only clear change listener if there are no more listening clients
+    if (clients.length === 0) {
+      packagerServer.setHMRFileChangeListener(null);
+    }
+  }
+
+  function connectedPlatformClients () {
+    return {
+      android: clients.filter(client => client.platform === "android"),
+      ios: clients.filter(client => client.platform === "ios"),
+    }
+  }
+
+  function forEachPlatform (callback) {
+    const platformClients = connectedPlatformClients();
+    return Promise.all([
+      wrapPotentialPromise(platformClients.android.length > 0 ? callback("android", platformClients.android) : null),
+      wrapPotentialPromise(platformClients.ios.length > 0 ? callback("ios", platformClients.ios) : null),
+    ]).then(results => ({
+       android: results[0],
+       ios: results[1],
+    }));
+  }
+
+  function forEachBundleEntry (callback) {
+    return forEachPlatform(platform =>
+      forEachKey(platformCache[platform], bundleEntry =>
+        callback(bundleEntry, platform)
+      )
+    )
+  }
+
+  function sendToClients (message, platform, bundleEntry) {
+    if (platform && bundleEntry) {
+      clientsPerBundleEntry[bundleEntry].map(client => {
+        if (client.platform === platform) {
+          client.ws.send(JSON.stringify(message))
+        }
+      })
+    }
+    else {
+      clients.forEach(client => client.ws.send(JSON.stringify(message)));
+    }
+  }
+
+  // Runs whenever a file changes
+  function onHMRChange (filename, stat)  {
+    if (clients.length === 0) {
+      return;
+    }
+
+    console.log(`[Hot Module Replacement] File change detected (${time()})`);
+    sendToClients({type: 'update-start'});
+
+    stat.then(() => {
+      return packagerServer.getShallowDependencies(filename)
+        .then(deps => {
+          if (clients.length === 0) {
+            return;
+          }
+
+          //Check the depedencies for each bundle entry
+          return forEachBundleEntry((bundleEntry, platform) => {
```

Comment:
I like how the code results when using this function, it makes iterating over the clients easy!. But couldn't you just iterate over the clients and use the `platform` and `bundleEntry` values stored on the objects?